protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,554 @@
+                            internal_addr: addr,
+                            multiaddr: multiaddr.clone(),
+                        };
+
+                        let duration = self.config.temporary.then_some(MAPPING_DURATION);
+                        if let Err(err) = gateway
+                            .sender
+                            .try_send(GatewayRequest::AddMapping(mapping.clone(), duration))
+                        {
+                            log::debug!(
+                                "could not request port mapping for {} on the gateway: {}",
+                                mapping.multiaddr,
+                                err
+                            );
+                        }
+
+                        self.mappings.insert(mapping, MappingState::Pending);
+                    }
+                    GatewayState::GatewayNotFound => {
+                        log::debug!(
+                            "network gateway not found, UPnP port mapping of {multiaddr} discarded"
+                        );
+                    }
+                };
+            }
+            FromSwarm::ExpiredListenAddr(ExpiredListenAddr {
+                listener_id,
+                addr: _addr,
+            }) => {
+                if let GatewayState::Available(ref mut gateway) = &mut self.state {
+                    if let Some((mapping, _state)) = self.mappings.remove_entry(&listener_id) {
+                        if let Err(err) = gateway
+                            .sender
+                            .try_send(GatewayRequest::RemoveMapping(mapping.clone()))
+                        {
+                            log::debug!(
+                                "could not request port removal for {} on the gateway: {}",
+                                mapping.multiaddr,
+                                err
+                            );
+                        }
+                        self.mappings.insert(mapping, MappingState::Pending);
+                    }
+                }
+            }
+            FromSwarm::ConnectionEstablished(_)
+            | FromSwarm::ConnectionClosed(_)
+            | FromSwarm::AddressChange(_)
+            | FromSwarm::DialFailure(_)
+            | FromSwarm::ListenFailure(_)
+            | FromSwarm::NewListener(_)
+            | FromSwarm::ListenerError(_)
+            | FromSwarm::ListenerClosed(_)
+            | FromSwarm::NewExternalAddrCandidate(_)
+            | FromSwarm::ExternalAddrConfirmed(_)
+            | FromSwarm::ExternalAddrExpired(_) => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: PeerId,
+        _connection_id: ConnectionId,
+        event: libp2p_swarm::THandlerOutEvent<Self>,
+    ) {
+        void::unreachable(event)
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        _params: &mut impl PollParameters,
+    ) -> Poll<ToSwarm<Self::ToSwarm, libp2p_swarm::THandlerInEvent<Self>>> {
```

Comment:
Nesting in this method is deep, making it difficult to reason about the control flow (e.g. the `loop` and the `return Poll::Pending` in nested arms). Do you have ideas how to simplify it or break it apart?

Reply:
This looks a lot better. Thanks @jxs.